File: typescript/problem_19.ts
// minCostPaint returns the minimum cost to paint N houses with K colors
// such that no two adjacent houses share the same color.
// Uses bottom-up DP. This runs in O(N*K) time and O(N*K) space.
export function minCostPaint(costs: number[][]): number {
    if (costs.length === 0) return 0;
    const n = costs.length, k = costs[0].length;
    const dp = costs.map(row => [...row]);

    for (let i = 1; i < n; i++)
        for (let c = 0; c < k; c++) {
            let minPrev = Infinity;
            for (let p = 0; p < k; p++)
                if (p !== c && dp[i-1][p] < minPrev) minPrev = dp[i-1][p];
            dp[i][c] += minPrev;
        }

    return Math.min(...dp[n - 1]);
}

export function problem19(): void {
    console.log("Problem 19:");
    const costs = [[17, 2, 17], [16, 16, 5], [14, 3, 19]];
    if (minCostPaint(costs)       !== 10) throw new Error("assertion failed");
    if (minCostPaint([[1, 2, 3]]) !== 1)  throw new Error("assertion failed");
    console.log("All Tests Passed");
}

File: rust/problem_19.rs
// min_cost_paint returns the minimum cost to paint N houses with K colors
// such that no two adjacent houses share the same color.
// Uses bottom-up DP. This runs in O(N*K) time.
pub fn min_cost_paint(costs: &[Vec<i64>]) -> i64 {
    if costs.is_empty() { return 0; }
    let n = costs.len();
    let k = costs[0].len();
    let mut dp: Vec<Vec<i64>> = costs.to_vec();

    for i in 1..n {
        for c in 0..k {
            let min_prev = (0..k).filter(|&p| p != c).map(|p| dp[i-1][p]).min().unwrap();
            dp[i][c] += min_prev;
        }
    }
    *dp[n-1].iter().min().unwrap()
}

pub fn problem19() {
    println!("Problem 19:");
    let costs = vec![vec![17,2,17], vec![16,16,5], vec![14,3,19]];
    assert_eq!(min_cost_paint(&costs),                10, "assertion failed");
    assert_eq!(min_cost_paint(&[vec![1,2,3]]),         1, "assertion failed");
    println!("All Tests Passed");
}
